src/future/batch_semaphore.rs

```diff
@@ -447,6 +492,15 @@ impl BatchSemaphore {
 
         state.waiters.remove(index).unwrap();
         assert!(waiter.is_queued.swap(false, Ordering::SeqCst));
+
+        if matches!(self.fairness, Fairness::StrictlyFair) && index == 0 {
+            // If the semaphore is strictly fair, and we removed the first waiter, check if its
+            // removal unblocks remaining waiters.  This can happen in the following situation:
```

Comment:
Nit-ish: When I read "This can happen in the following situation", I usually expect the situation abstractly, and not an example.

Also: Example is bigger than it needs to be. If you have already decided that W1 is going to attempt to acquire an unacquireable amount, then we can make do with 1 permit, W1 wanting to acq 2, and W2 wanting to acq 1, then dropping W1.

Off-topic: do we have checks that we never try to acquire more than the maximum number of permits?

Reply:
I tried writing it abstractly, but found a concrete example was a lot easier to read.  Also, the shortest example requires weird behavior from W1 - it's woken up but then decides to drop.  I felt the example I have is more realistic (W2 gives up after waiting, e.g. with a timeout).

Also: I don't understand what you mean by `maximum number of permits`?  We can initialize a BatchSemaphore with 0 permits and then someone can release thousands of permits later on. (We use this for mpsc channels for instance.)

Reply:
> tried writing it abstractly, but found a concrete example was a lot easier to read. 

Yes, I like the example, just more of a nit on writing "An example of this" or similar over "This can happen in the following situation".

> weird behavior from W1 - it's woken up but then decides to drop.

? I don't follow. The example is the same, but W1 wants to acquire 2 permits (with 1 available), W2 wants to acquire 1 permits, and W3 doesn't exist. Rest of the example pans out the same.

> Also: I don't understand ...

Yeah, nevermind.